PP-484:PP-485:Tests for qstat dsv and json output

visheshh opened the pull request:
PP-484:PP-485:Tests for qstat dsv and json output
PTL tests for qstat -f -F json and qstat -f -F dsv

Issue-ID

PP-484
PP-485
Problem description

Test cases required for testing qstat dsv and json output
Solution description

Added test cases for qstat -f -F json , qstat -f -F dsv , qstat -f -F dsv -D, qstat -Bf -F json , qstat -Bf -F dsv , qstat -Bf -F dsv -D, qstat -Qf -F json , qstat -Qf -F dsv , qstat -Qf -F dsv -D


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

visheshh commented:
I have added smoke tag to the tests to make it run on Travis CI , Will remove once the tests are reviewed .

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        jid = self.server.submit(j)
+        f = open('/tmp/qstat_script', 'w')
+        f.write(self.server.pbs_conf['PBS_EXEC'] +
+                "/bin/qstat -f -F json > /tmp/qstat_out")
```
Please use os.path.join() like you did in the dsv test.

visheshh replied:
Done

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        os.chmod("/tmp/qstat_script", 0755)
+        run_script = "sh /tmp/qstat_script"
+        ret = self.du.run_cmd(self.server.hostname, cmd=run_script, sudo=True)
+        print ret['out']
```
please remove this print statement.

visheshh replied:
Done

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        print ret['out']
+        f = open('/tmp/qstat_out', 'r')
+        qstat_out = f.read()
+        print qstat_out
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        qstat_attr = []
+        for key, val in json_object.iteritems():
+            qstat_attr.append(str(key))
+            if type(val) == type(json_object):
```
as above consider using isinstance()

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        ret = self.du.run_cmd(self.server.hostname, cmd=run_script, sudo=True)
+        print ret['out']
+        f = open('/tmp/qstat_out', 'r')
+        qstat_out = f.read()
```
Why jump through the hoop of writing the output out to a file where you didn't in the dsv tests?  You read those directly out of ret['out'].

visheshh replied:
ret['out'] breaks json output line by line and makes it an array of lines . Which makes it unusable for json modules.

bhroam replied:
You could always recombine the array with a "\n".join().  I still like the idea of the test not assuming a list of attributes.  It'll go stale.  If it's still too much of a pain to recombine it for the json modules, I understand.

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        run_script = "sh /tmp/qstat_script"
+        ret = self.du.run_cmd(self.server.hostname,
+                              cmd=run_script, runas=TEST_USER)
+        print ret['out']
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        print ret['out']
+        f = open('/tmp/qstat_out', 'r')
+        qstat_out = f.read()
+        print qstat_out
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            json_object = json.loads(qstat_out)
+        except ValueError, e:
+            self.assertTrue(False)
+        self.assertTrue(True)
```
This is a noop.  You're asserting True is True.

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        qstat_attr = []
+        for key, val in json_object.iteritems():
+            qstat_attr.append(str(key))
+            if type(val) == type(json_object):
```
consider using isinstance()

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            if attr not in qstat_attr:
+                print attr + "is missing"
+                self.assertTrue(False)
+        self.assertTrue(True)
```
This is a noop.

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        for attr in attrs_qstatf:
+            if attr not in qstat_attr:
+                print attr + "is missing"
+                self.assertTrue(False)
```
Turn this into a self.assertTrue() with a failure message.

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        """
+        j = Job(TEST_USER)
+        j.set_sleep_time(10)
+        j.set_attributes({ATTR_J: '1-3:1'})
```
no need for the :1

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        os.chmod("/tmp/qstat_script", 0755)
+        run_script = "sh /tmp/qstat_script"
+        ret = self.du.run_cmd(self.server.hostname, cmd=run_script, sudo=True)
+        print ret['out']
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        print ret['out']
+        f = open('/tmp/qstat_out', 'r')
+        qstat_out = f.read()
+        print qstat_out
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            json_object = json.loads(qstat_out)
+        except ValueError, e:
+            self.assertTrue(False)
+        self.assertTrue(True)
```
This is a noop

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        qstat_attr = []
+        for key, val in json_object.iteritems():
+            qstat_attr.append(str(key))
+            if type(val) == type(json_object):
```
consider using isinstance()

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+                qstat_attr.append(qstat_attrs)
+        print qstat_attr
+        for attr in attrs_qstatf:
+            if attr not in qstat_attr:
```
Turn this into a self.assertTrue() with a failure message

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            if attr not in qstat_attr:
+                print attr + "is missing"
+                self.assertTrue(False)
+        self.assertTrue(True)
```
This is a noop

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        os.chmod("/tmp/qstat_script", 0755)
+        run_script = "sh /tmp/qstat_script"
+        ret = self.du.run_cmd(self.server.hostname, cmd=run_script, sudo=True)
+        print ret['out']
```
remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        print ret['out']
+        f = open('/tmp/qstat_out', 'r')
+        qstat_out = f.read()
+        print qstat_out
```
remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            json_object = json.loads(qstat_out)
+        except ValueError, e:
+            self.assertTrue(False)
+        self.assertTrue(True)
```
this is a noop

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+                qstat_attr.append(qstat_attrs)
+        print qstat_attr
+        for attr in attrs_qstatbf:
+            if attr not in qstat_attr:
```
Turn this into a self.assertTrue() with a failure message

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            if attr not in qstat_attr:
+                print attr + "is missing"
+                self.assertTrue(False)
+        self.assertTrue(True)
```
this is a noop

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        os.chmod("/tmp/qstat_script", 0755)
+        run_script = "sh /tmp/qstat_script"
+        ret = self.du.run_cmd(self.server.hostname, cmd=run_script, sudo=True)
+        print ret['out']
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        print ret['out']
+        f = open('/tmp/qstat_out', 'r')
+        qstat_out = f.read()
+        print qstat_out
```
please remove this print statement

visheshh replied:
Done

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            json_object = json.loads(qstat_out)
+        except ValueError, e:
+            self.assertTrue(False)
+        self.assertTrue(True)
```
this is a noop

visheshh replied:
removed

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        qstat_attr = []
+        for key, val in json_object.iteritems():
+            qstat_attr.append(str(key))
+            if type(val) == type(json_object):
```
consider using isinstance()

visheshh replied:
Done

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            if type(val) == type(json_object):
+                qstat_attrs = self.parse_json(val, qstat_attr)
+                qstat_attr.append(qstat_attrs)
+        print qstat_attr
```
please remove this print statement

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+                qstat_attr.append(qstat_attrs)
+        print qstat_attr
+        for attr in attrs_qstatqf:
+            if attr not in qstat_attr:
```
Turn this into a self.assertTrue() with a failure message

visheshh replied:
Done

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+            if attr not in qstat_attr:
+                print attr + "is missing"
+                self.assertTrue(False)
+        self.assertTrue(True)
```
this is a noop

visheshh replied:
removed

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+            'Resource_List.place', 'Resource_List.select', 'schedselect',
+            'substate', 'Variable_List', 'euser', 'egroup', 'queue_rank',
+            'queue_type', 'etime', 'Submit_arguments', 'executable',
+            'argument_list', 'project']
```
Thanks for the comment . I will implement this for dsv . For json the format differs , So i will leave the static values as it is.
Json : {
    "timestamp":1492166697,
    "pbs_version":"17.2.0.20170330010525",
    "pbs_server":"localhost",
    "Jobs":{
        "109.localhost":{
  qstat -f : Job Id: 109.localhost
    Job_Name = STDIN

visheshh commented:
Hi Bhroam,
Thanks for the review.
I have addressed most of the comments .
The formatting in json is pretty different and getting the same list from qstat -f is not possible .
For ex: qstat -f gives Job ID : <109.localhost> and qstat -f -F json gives Jobs : { 109.localhost , there are some other variables in json like timestamp .
I have generated the list on the fly for dsv tests.
If we take output of the command by not writing to a file , The output is already broken into lines . So it becomes unreable by json modules .

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+
+
+@tags('smoke')
+class Qstat_dsv(PBSTestSuite):
```
Done

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        elif qstat_type == "queue":
+            qstat_cmd = os.path.join(self.server.pbs_conf[
+                'PBS_EXEC'], 'bin', 'qstat') + ' -Qf -F dsv ' + delim
+        ret = self.du.run_cmd(self.server.hostname, cmd=qstat_cmd, sudo=True)
```
If test asks to test for super user , I have left sudo=True else removed .

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        for item in attr_vals:
+            qstat_attr = item.split("=")
+            qstat_attrs.append(qstat_attr[0])
+        print qstat_attrs
```
Done

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        j = Job(TEST_USER)
+        j.set_sleep_time(10)
+        jid = self.server.submit(j)
+        print jid
```
Done

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        print jid
+        qstat_cmd = os.path.join(self.server.pbs_conf[
+            'PBS_EXEC'], 'bin', 'qstat') + ' -f -F dsv ' + str(jid)
+        ret = self.du.run_cmd(self.server.hostname, cmd=qstat_cmd, sudo=True)
```
If test asks to test for super user , I have left sudo=True else removed

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        j = Job(TEST_USER)
+        j.set_sleep_time(10)
+        jid = self.server.submit(j)
+        print jid
```
Done

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        """
+        j = Job(TEST_USER)
+        j.set_sleep_time(10)
+        j.set_attributes({ATTR_J: '1-3:1'})
```
Done

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        j.set_sleep_time(10)
+        j.set_attributes({ATTR_J: '1-3:1'})
+        jid = self.server.submit(j)
+        print jid
```
Done

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -0,0 +1,259 @@
+        j = Job(TEST_USER)
+        j.set_sleep_time(10)
+        jid = self.server.submit(j)
+        print jid
```
Done

visheshh commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -0,0 +1,292 @@
+        """
+        for key, val in dict.items():
+            qstat_attr.append(str(key))
+            if type(val) == type(dict):
```
Done

bhroam commented on test/tests/functional/pbs_qstat_json.py:
```diff
@@ -67,22 +67,20 @@ def test_qstat_json_valid(self):
-        f.write(self.server.pbs_conf['PBS_EXEC'] +
-                "/bin/qstat -f -F json > /tmp/qstat_out")
+        f.write(os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                             'bin', 'qstat') + " -f -F json > /tmp/qstat_out")
```
While it is highly unlikely, if two copies of this test were being run at the same time, you'd get a race condition.  One test's file would overwrite the other.  Maybe use tempfile.mkstemp()?

visheshh replied:
I will use "\n".join it makes sense rather than using tempfile . Thanks for the comment

visheshh commented on test/tests/functional/pbs_qstat_dsv.py:
```diff
@@ -55,15 +55,28 @@ def parse_dsv(self, attrs_qstatf, jid, qstat_type, delimiter=None):
+                                         'bin', 'qstat') + cmd
             qstat_cmd = os.path.join(self.server.pbs_conf['PBS_EXEC'],
-                                     'bin', 'qstat') + cmd
+                                     'bin', 'qstat') + ' -f ' + str(jid)
```
If i do PTL status call and take the keys , PTL formats 'Job Id: <job id>' into 'id' . If i send that list , test will fail as id is missing . I think the current approach is better , as it tests the real command outputs .

bhroam replied:
Good point.  I had forgotten about the change of keys.  Keeping as direct calls to qstat seems best.

visheshh commented:
Hi @bhroam , Addressed review comments , Please review .

bhroam commented:
Thanks for making all the changes.  It looks good to me.  I sign off.

visheshh commented:
Squashed and rebased the branch .

nithinj commented:
Can we have a test case where you submit the job using `-V` option which will send the environment also to the job? Since the environment variables usually contain a lot of special characters, which will be a good way to confirm that it does not break our encoding. You can either add a new test case or add `-V` to one of the existing test cases.
